### ISSUE TITLE
Feature Request: CSV Upload for Bulk Search/Replace Operations

### ISSUE THREAD

**Issue by BOTSONS** [ExpandedFronts/Better-Search-Replace#132]
## Summary

Add CSV file upload capability to enable users to process multiple search/replace pairs in a single operation.

## Problem Statement

Currently, users who need to perform multiple search/replace operations must:
1. Run the tool once for each pair
2. Manually enter each search/replace value
3. Wait for each operation to complete
4. Risk forgetting values or making input errors

This becomes particularly painful when:
- Migrating sites with multiple domain/URL patterns to update
- Performing site-wide content updates (emails, phone numbers, addresses)
- Moving from staging to production with environment-specific values
- Rebranding projects requiring dozens of text replacements

## Proposed Solution

Allow users to upload a CSV file containing multiple search/replace pairs that are processed sequentially across selected tables.

### CSV Format
```csv
search_for,replace_with
http://old-site.com,https://new-site.com
old_email@example.com,new_email@example.com
555-OLD-PHON,555-NEW-PHON
```

### For Users
- **Time Savings**: Process dozens of replacements in one operation instead of many
- **Accuracy**: No manual re-typing between operations
- **Repeatability**: Save CSV files for future use or documentation
- **Transparency**: See all changes in one comprehensive results report
- **Safety**: Dry run mode works across all CSV pairs

### For the Plugin
- **Competitive Advantage**: Few WordPress search/replace plugins offer this
- **User Satisfaction**: Addresses a common pain point
- **Professional Use Cases**: Appeals to agencies and developers

### 1. Site Migration
Moving WordPress from staging to production:
```csv
search_for,replace_with
http://staging.clientsite.com,https://www.clientsite.com
staging@clientsite.com,info@clientsite.com
/var/www/staging,/var/www/production
```

### 2. Multi-Domain Setup
Updating references across a network:
```csv
search_for,replace_with
old-cdn.example.com,new-cdn.example.com
old-images.example.com,new-images.example.com
old-assets.example.com,new-assets.example.com
```

### 3. Rebranding
Company name and contact changes:
```csv
search_for,replace_with
Old Company Inc,New Brand LLC
oldcompany.com,newbrand.com
contact@oldcompany.com,hello@newbrand.com
```

## Implementation Details

I've implemented a working prototype with the following features:

### User Interface
- CSV file upload field with file type validation
- Live preview showing first 10 rows of uploaded CSV
- Automatic header detection (skips rows with "search_for" header)
- Disables manual input fields when CSV is uploaded
- Clear error messages for invalid files

### Processing
- Sequential processing: Each CSV pair runs across all selected tables before moving to next pair
- Progress tracking with real-time updates
- Supports dry run mode (test all pairs without making changes)
- Handles serialized WordPress data correctly
- Maximum file size: 5MB (configurable)

### Results Display
- Overall summary (X pairs processed, Y total changes)
- Individual results for each CSV pair showing:
  - What was searched and replaced
  - Number of changes per pair
  - Table-by-table breakdown
  - Processing time

### Security
- WordPress nonce verification
- File type validation (CSV only)
- File size limits
- Uploaded file validation via `is_uploaded_file()`
- Transient-based temporary storage with auto-expiration
- Transient key validation with regex pattern matching
- Respects existing `bsr_capability` filter

### Code Quality
- Follows WordPress Coding Standards
- Yoda conditions throughout
- Comprehensive inline documentation
- All functions documented with @since, @access, @param, @return tags
- Tab indentation (size 4)
- No breaking changes to existing functionality

### New Files
- `includes/class-bsr-csv.php` - CSV parsing, validation, and storage logic

### Modified Files
- `includes/class-bsr-main.php` - Load CSV class
- `includes/class-bsr-admin.php` - Enhanced results display for CSV operations
- `includes/class-bsr-ajax.php` - CSV upload handler and multi-pair processing
- `templates/bsr-search-replace.php` - CSV upload UI elements
- `assets/js/better-search-replace.js` - Client-side CSV handling and preview

### No Database Changes
- Uses WordPress transients for temporary storage
- No new tables or schema modifications required

## Testing Performed

- Upload valid CSV with multiple pairs → Processes successfully
- Upload non-CSV file → Rejects with clear error
- Upload CSV with invalid format → Shows helpful error
- Dry run with CSV → Previews all changes without modifying database
- CSV with header row → Automatically detects and skips
- CSV without header → Processes all rows correctly
- CSV with special characters → Handles correctly
- Empty CSV → Rejects gracefully
- Large CSV (50+ pairs) → Processes without timeout

## Backward Compatibility

- No breaking changes
- Single search/replace works exactly as before
- Optional feature - can be completely ignored by users who don't need it
- Respects all existing settings and permissions

### 1. Free vs. Pro Version
This feature provides significant time savings for users. I notice the Pro version advertises "time-saving features" and includes the ability to "save or load custom profiles."

**Would you prefer this feature to be:**
- [ ] Part of the free version (increases free version appeal)
- [ ] Part of the Pro version (additional Pro selling point)
- [ ] Split implementation (basic CSV in free, advanced features in Pro)

### 2. Feature Scope
**Are there any modifications you'd like to see?**
- Different CSV format?
- Additional validation?
- Export results to CSV?
- Ability to save frequently-used CSV files?

### 3. Pull Request
If you're interested in this feature, I'm happy to:
- Submit a pull request with the current implementation
- Make any requested changes
- Add unit tests
- Update documentation

## Additional Resources

I've prepared:
- Full implementation with WordPress Coding Standards compliance
- Comprehensive feature documentation
- Example CSV files with usage guide
- Inline code documentation

**I'm ready to contribute this to the project if there's interest. Please let me know your thoughts on:**
1. Whether this aligns with the plugin's roadmap
2. Free vs. Pro placement
3. Any changes or improvements you'd like to see

Thank you for considering this feature request!

---

**Labels**: enhancement, feature-request
**Version**: Works with current develop branch
**WordPress**: Tested on WordPress 6.4+
**PHP**: Compatible with PHP 7.4+

**Comment by L1QU2D** [comment]
+1 for this
